index.js

```diff
@@ -265,21 +276,51 @@ FacebookInsightStream.prototype._collect = function ( metrics, item, buffer, eve
                 }
             }
         })
+        .catch( SkipedError, function ( error ) {
+            console.log( "facebook-insights skiped error", error );    
+        })
         .then( function () {
+            // remove the current paramater when done
+            var _metricIdx = metrics.indexOf( _metric );
+            var _evIdx = events.indexOf( _ev );
+            metrics.splice( _metricIdx, 1 );
+            events.splice( _evIdx, 1 );
+
             return this._collect( metrics, item, buffer, events );
         })
         .catch( function ( error ) {
-            if ( error.skip ) {
-                return this._collect( metrics, item, buffer )
-            } else {
-                this.emit( "error", error )
-            }
+            var retry = this._collect.bind( this, metrics, item, buffer, events );
+            return this.handleError( error, retry );
         })
 }
 
+/**
+ * Thrown error handling methos, when any part of the stream throws an error
+ * it pass its error to this method, with retry function and the thrown error
+ * this function decides if this error is retryable, and then retry the method
+ * that generated the error by colling to retry, otherwise it emmits error.
+ *
+ * Overide this method to create your own error handling and retrying mechanism
+ * 
+ * @param  {Error}  error 
+ * @param  {Function} retry the function that should be invoke to retry the process
+ */
+FacebookInsightStream.prototype.handleError = function ( error, retry ) {
+    if ( error.retry === true ) {
+        return retry();
+    } else {
+        this.emit( 'error', error );
+    }
+}
+
+// predicate-based error filter 
+function SkipedError ( error ) {
+    return error.skip === true;
+}
+
 function errorHandler ( body )  {
     if ( body.error ) {
-        throw new Error( body.error.message )
+        throw body.error
```

Comment:
@liorrozen @alonweissfeld Facebook's API return reach description for the error, i want to keep the full error and not just the message.